python/ray/rllib/ppo/loss.py

```diff
@@ -38,50 +34,36 @@ def __init__(
             # do not make the standard deviations free variables.
             vf_config["free_log_std"] = False
             with tf.variable_scope("value_function"):
+                if ADB:
+                    input_vf = tf.concat([self.observations, self.actions], 1)
+                else:
+                    input_vf = self.observations
                 self.value_function = ModelCatalog.get_model(
-                    registry, observations, 1, vf_config).outputs
+                    registry, input_vf, 1, vf_config).outputs
             self.value_function = tf.reshape(self.value_function, [-1])
 
-        curr_logp = self.curr_dist.logp(actions)
-        prev_logp = self.prev_dist.logp(actions)
+        # Make loss functions.
+        if ADB:
+            self.ratio = tf.div(self.curr_dist.r_matrix(actions),
+                            self.prev_dist.r_matrix(actions))
+        else:
+            self.ratio = tf.exp(self.curr_dist.logp(actions) -
+                                self.prev_dist.logp(actions))
         self.kl = self.prev_dist.kl(self.curr_dist)
+        self.mean_kl = tf.reduce_mean(self.kl)
         self.entropy = self.curr_dist.entropy()
-
-        # handle everything uniform as if it were the multiagent case
-        if not isinstance(curr_logp, list):
-            self.kl = [self.kl]
-            curr_logp = [curr_logp]
-            prev_logp = [prev_logp]
-            self.entropy = [self.entropy]
-
-        # Make loss functions.
-        self.ratio = [tf.exp(curr - prev)
-                      for curr, prev in zip(curr_logp, prev_logp)]
-        self.mean_kl = [tf.reduce_mean(kl_i) for kl_i in self.kl]
         self.mean_entropy = tf.reduce_mean(self.entropy)
-        self.surr1 = [ratio_i * advantages for ratio_i in self.ratio]
-        self.surr2 = [tf.clip_by_value(ratio_i, 1 - config["clip_param"],
-                                       1 + config["clip_param"]) * advantages
-                      for ratio_i in self.ratio]
-        self.surr = [tf.minimum(surr1_i, surr2_i) for surr1_i, surr2_i in
-                     zip(self.surr1, self.surr2)]
-        self.surr = tf.add_n(self.surr)
-        self.mean_policy_loss = tf.reduce_mean(-self.surr)
-
-        self.entropy = tf.add_n(self.entropy)
-        entropy_prod = config["entropy_coeff"]*self.entropy
-
-        # there's only one kl value for a shared model
-        if self.shared_model:
-            kl_prod = tf.add_n([kl_coeff[0] * kl_i for
-                                i, kl_i in enumerate(self.kl)])
-            # all the above values have been rescaled by num_agents
-            self.surr /= self.num_agents
-            kl_prod /= self.num_agents
-            entropy_prod /= self.num_agents
+        self.surr1 = self.ratio * advantages
+        if ADB:
+            action_dim = action_space.shape[0]
+            self.surr2 = tf.clip_by_value(self.ratio, (1 - config["clip_param"])**(1/action_dim),
```

Comment:
Are these ranges reasonable to you?

For m=3 actions, `clip_param`=0.2:
(0.928, 1.063)

For m=8 actions, `clip_param`=0.2:
(0.972, 1.023)